File: MyIntent/app/src/main/java/com/example/intent/MainActivity.java
package com.example.intent;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.content.*;
import android.view.*;
import android.widget.*;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {
    public static final int REQUEST_CODE_MENU = 101;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button button = findViewById(R.id.button);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) { // 클릭 시 Main -> Menu 전환
                // Context는 버튼이면 버튼이 포함되어 있는 주변 정보를 갖고 있음(레이아웃, 환경 등)
                // 안드로이드에서 UI 객체는 생성시 Context 넘겨받도록 되어 있음.
                // OnCreate 함수 안에서는 this가 Activity 객체가 되므로 버튼 등의 Context이다.

                // MenuActivity에 대한 참조를 가능하게 함.
                Intent intent = new Intent(getApplicationContext(), MenuActivity.class);
                startActivityForResult(intent, REQUEST_CODE_MENU); // 어떤 화면으로부터 응답을 받은 건지 구분 가능한 코드
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(requestCode == REQUEST_CODE_MENU) { // MenuActivity에서 보내는 코드
            Toast.makeText(getApplicationContext(), "OnActivityResult 호출\n 요청 코드: " + requestCode
                    + "\n결과 코드:" + resultCode, Toast.LENGTH_LONG).show();
            if(data != null) {
                String name = data.getStringExtra("name");
                if(name != null) {
                    Toast.makeText(this, "응답으로 받은 데이터: " + name, Toast.LENGTH_LONG).show();
                }
            }
        }
    }
}

File: MyKakaoMsg/app/src/main/java/com/example/mykakaomsg/MainActivity.kt
package com.example.mykakaomsg

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.TextView
import com.kakao.sdk.common.util.KakaoCustomTabsClient
import com.kakao.sdk.link.LinkClient
import com.kakao.sdk.link.WebSharerClient
import com.kakao.sdk.talk.TalkApiClient
import com.kakao.sdk.template.model.DefaultTemplate

class MainActivity : AppCompatActivity() {
    private final var TAG = "MainActivity"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        var intent = intent

        if (LinkClient.instance.isKakaoLinkAvailable(this)) {
            // 카카오톡으로 공유
            LinkClient.instance.scrapTemplate(this, "https://example.com/msgcallback") { linkResult, error ->
                if (error != null) {
                    Log.e(TAG, "카카오링크 실패", error)
                } else {
                    startActivity(linkResult!!.intent)
                }
            }
        } else {
            // 웹으로 공유
            WebSharerClient.instance.scrapTemplateUri(	"https://example.com/msgcallback").let {

                KakaoCustomTabsClient.openWithDefault(this, it)
                // 또는
                //startActivity(Intent(Intent.ACTION_VIEW, it))
            }
        }

//        val serverCallbackArgs = mapOf("templateId" to "46294")
//
//        val memo = LinkClient.instance.customTemplate(this, 46294, serverCallbackArgs) { linkResult, error ->
//            if (error != null) {
//                Log.e(TAG, "카카오링크 보내기 실패", error)
//            }
//            else if (linkResult != null) {
//                Log.d(TAG, "카카오링크 보내기 성공 ${linkResult.intent}")
//                startActivity(linkResult.intent)
//
//                // 카카오링크 보내기에 성공했지만 아래 경고 메시지가 존재할 경우 일부 컨텐츠가 정상 동작하지 않을 수 있습니다.
//                Log.w(TAG, "Warning Msg: ${linkResult.warningMsg}")
//                Log.w(TAG, "Argument Msg: ${linkResult.argumentMsg}")
//            }
//        }
//
//        var button = findViewById(R.id.Button) as Button
//        button.setOnClickListener(View.OnClickListener {
//            Log.d(TAG, "clicked")
//            TalkApiClient.instance.sendDefaultMemo(memo) { error ->
//                if (error != null) {
//                    Log.e(TAG, "나에게 보내기 실패", error)
//                } else {
//                    Log.i(TAG, "나에게 보내기 성공")
//                }
//            }

//            LinkClient.instance.customTemplate(context = this, templateId = 46294) { linkResult, error ->
//                if (error != null) {
//                    Log.e(TAG, "카카오링크 보내기 실패", error)
//                }
//                else if (linkResult != null) {
//                    Log.d(TAG, "카카오링크 보내기 성공 ${linkResult.intent}")
//                    startActivity(linkResult.intent)
//
//                    // 카카오링크 보내기에 성공했지만 아래 경고 메시지가 존재할 경우 일부 컨텐츠가 정상 동작하지 않을 수 있습니다.
//                    Log.w(TAG, "Warning Msg: ${linkResult.warningMsg}")
//                    Log.w(TAG, "Argument Msg: ${linkResult.argumentMsg}")
//                }
//            }
//        })
    }

}

private fun TalkApiClient.sendDefaultMemo(template: Unit, callback: (Throwable?) -> Unit) {

}
